src/langserver/debugger/auxtools.rs

```diff
@@ -0,0 +1,432 @@
+        }
+    }
+
+    pub fn get_stacks(&mut self) -> Result<Vec<Stack>, Box<dyn std::error::Error>> {
+        self.send_or_disconnect(Request::Stacks)?;
+
+        match self.read_response_or_disconnect()? {
+            Response::Stacks { stacks } => Ok(stacks),
+            _ => Err(Box::new(super::GenericError("received incorrect response"))),
+        }
+    }
+
+    pub fn get_stack_frames(
+        &mut self,
+        stack_id: u32,
+        start_frame: Option<u32>,
+        count: Option<u32>,
+    ) -> Result<(Vec<StackFrame>, u32), Box<dyn std::error::Error>> {
+        self.send_or_disconnect(Request::StackFrames {
+            stack_id,
+            start_frame,
+            count,
+        })?;
+
+        match self.read_response_or_disconnect()? {
+            Response::StackFrames {
+                frames,
+                total_count,
+            } => Ok((frames, total_count)),
+            _ => Err(Box::new(super::GenericError("received incorrect response"))),
+        }
+    }
+
+    // TODO: return all the scopes
+    pub fn get_scopes(&mut self, frame_id: u32) -> Result<(Option<VariablesRef>, Option<VariablesRef>, Option<VariablesRef>), Box<dyn std::error::Error>> {
+        self.send_or_disconnect(Request::Scopes {
+            frame_id
+        })?;
+
+        match self.read_response_or_disconnect()? {
+            Response::Scopes {
+                arguments,
+                locals,
+                globals,
+            } => Ok((arguments, locals, globals)),
+            _ => Err(Box::new(super::GenericError("received incorrect response"))),
+        }
+    }
+
+    pub fn get_variables(&mut self, vars: VariablesRef) -> Result<Vec<Variable>, Box<dyn std::error::Error>> {
+        self.send_or_disconnect(Request::Variables {
+            vars
+        })?;
+
+        match self.read_response_or_disconnect()? {
+            Response::Variables { vars } => Ok(vars),
+            _ => Err(Box::new(super::GenericError("received incorrect response"))),
+        }
+    }
+
+    pub fn get_last_error_message(&self) -> String {
+        self.last_error.read().unwrap().clone()
+    }
+
+    pub fn set_catch_runtimes(&mut self, should_catch: bool) -> Result<(), Box<dyn std::error::Error>> {
+        self.send_or_disconnect(Request::CatchRuntimes { should_catch })
+    }
+}
+
+impl AuxtoolsThread {
+	fn spawn_listener(
+		self,
+		listener: TcpListener,
+		connection_sender: mpsc::Sender<TcpStream>,
+	) -> JoinHandle<()> {
+		thread::spawn(move || match listener.accept() {
+			Ok((stream, _)) => {
+				match connection_sender.send(stream.try_clone().unwrap()) {
+					Ok(_) => {}
+					Err(e) => {
+						eprintln!("Debug client thread failed to pass cloned TcpStream: {}", e);
+						return;
+					}
+                }
+                
+				self.run(stream);
+			}
+
+			Err(e) => {
+				eprintln!("Debug client failed to accept connection: {}", e);
+			}
+		})
+    }
```

Comment:
mixed indentation getting worse

Reply:
wouldn't be a problem if running `cargo fmt` didn't try to fix up every file in the repo ;_;